### PR TITLE
ipatests: Trust-functional selinux-usermap port part3.

### PR DESCRIPTION
**Pull request by pranav210798**
## Summary by Sourcery

Extend SELinux user map trust integration coverage with new precedence, hostgroup, and default mapping scenarios for trusted AD users, and improve test robustness and helpers around SSSD caching and SSH SELinux verification.

Enhancements:
- Add SELinux user constants for additional SELinux roles used across trust functional tests.
- Add helper methods to prime SSSD caches for trusted users and to run SSH-based SELinux verification from different hosts.
- Increase post-cache-reset wait time to allow SSSD to retrieve updated HBAC, SELinux map, and trust data and clarify an SSH helper docstring.

Tests:
- Add extensive SELinux user map integration tests covering HBAC-linked maps, hostgroup-scoped maps, multi-map precedence, empty and guest_u defaults, and their interaction with SSH access for forest-root and subdomain trusts.
- Strengthen existing SELinux user map tests by ensuring trusted users and group memberships resolve before group setup and by asserting external group membership correctness.

### CODE REVIEW

**Comment by rcritten** [comment]
The lint failure is a duplicated test name:
ipatests/test_integration/test_trust_functional.py:4929: [E0102(function-redefined), TestTrustFunctionalSelinuxUsermap.test_selinuxusermap_010_guest_default] method already defined line 4356)

But in looking at the naming the numbers are not incrementing, they restart at least once. Do we need the numbers at all? Is it there to enforce ordering?

File: ipatests/test_integration/test_trust_functional.py [hunk 1]
```diff
@@ -2747,6 +2753,8 @@ def _selinux_cache_reset_all(cls):
         for host in (cls.master, *cls.clients):
             tasks.clear_sssd_cache(host)
             tasks.wait_for_sssd_domain_status_online(host)
+        # give time to SSSD to retrieve new records (HBAC, selinux maps, trust)
+        time.sleep(30)
```
Comment (rcritten) [inline ipatests/test_integration/test_trust_functional.py]: Is there a reason 30 seconds was chosen (and not say 25)? This function is called at least 5 times so this adds 2 1/2 minutes of pure sleep.

File: ipatests/test_integration/test_trust_functional.py [hunk 2]
```diff
@@ -3942,6 +3969,627 @@ def test_selinuxusermap_006_hbac_hostgroup_map(self):
...
+                tasks.kinit_admin(self.master)
+                for _map in (
+                    'selinuxusermaprule1', 'selinuxusermaprule2',
+                    'selinuxusermaprule3',
+                ):
+                    tasks.selinuxusermap_del(
+                        self.master, _map, raiseonerr=False,
+                    )
+
+    def test_selinuxusermap_10_guest_default(self):
```
Comment (rjeffman) [inline ipatests/test_integration/test_trust_functional.py]: Other tests use `000` mask for naming, this one should also use it.

Comment (pranav210798) [reply ipatests/test_integration/test_trust_functional.py]: Sure

File: ipatests/test_integration/test_trust_functional.py [hunk 3]
```diff
@@ -3942,6 +3969,618 @@ def test_selinuxusermap_006_hbac_hostgroup_map(self):
...
+                    self.ad_user_password, expect_success=False,
+                )
+                ssh_with_password(
+                    self.master, user2, self.clients[0].hostname,
+                    self.ad_user_password, expect_success=False,
+                )
+                ssh_with_password(
+                    self.master, user2, self.clients[1].hostname,
+                    self.ad_user_password, expect_success=False,
+                )
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: This can be improved with : 
```
test_cases = [
    (user1, self.clients[0].hostname, True),
    (user1, self.clients[1].hostname, True),
    (user1, self.master.hostname, False),
    (user2, self.clients[0].hostname, False),
    (user2, self.clients[1].hostname, False),
]

for user, hostname, should_succeed in test_cases:
    ssh_with_password(
        self.master,
        user,
        hostname,
        self.ad_user_password,
        expect_success=should_succeed,
    )
```

File: ipatests/test_integration/test_trust_functional.py [hunk 4]
```diff
@@ -3942,6 +3969,618 @@ def test_selinuxusermap_006_hbac_hostgroup_map(self):
...
+                    [f'--selinuxuser={SELINUX_T4}'],
+                )
+                tasks.selinuxusermap_add_user(
+                    self.master, 'selinuxusermaprule3',
+                    groups=group1,
+                )
+                tasks.selinuxusermap_add_hosts(
+                    self.master, 'selinuxusermaprule3',
+                    hosts=self.clients[0].hostname,
+                )
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: Since the three rule blocks only differ by the rule name and SELinux user, you can loop over them.
```
rules = [
    ("selinuxusermaprule1", SELINUX_T3),
    ("selinuxusermaprule2", SELINUX_T2),
    ("selinuxusermaprule3", SELINUX_T4),
]

for rule_name, selinux_user in rules:
    tasks.selinuxusermap_add(
        self.master,
        rule_name,
        [f"--selinuxuser={selinux_user}"],
    )
    tasks.selinuxusermap_add_user(
        self.master,
        rule_name,
        groups=group1,
    )
    tasks.selinuxusermap_add_hosts(
        self.master,
        rule_name,
        hosts=self.clients[0].hostname,
    )
```

File: ipatests/test_integration/test_trust_functional.py [hunk 5]
```diff
@@ -3942,6 +3969,618 @@ def test_selinuxusermap_006_hbac_hostgroup_map(self):
...
+                self._kinit_trust_user(self.master, user2, domain)
+                verify_ssh_selinuxuser_with_krbcred(
+                    self.master, user2, self.clients[0], SELINUX_T3_VERIF,
+                    expect_match=False)
+                verify_ssh_selinuxuser_with_krbcred(
+                    self.master, user2, self.clients[0], SELINUX_DEFAULT_VERIF)
+                verify_ssh_selinuxuser_with_krbcred(
+                    self.master, user2, self.clients[1], SELINUX_DEFAULT_VERIF)
+                verify_ssh_selinuxuser_with_krbcred(
+                    self.master, user2, self.master, SELINUX_DEFAULT_VERIF)
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: This can improved with expected behavior for each user is explicitly listed, making the test cases easier to read and maintain.
```
test_cases = {
    user1: [
        (self.clients[0], SELINUX_T2_VERIF, False),
        (self.clients[0], SELINUX_T3_VERIF, True),
        (self.clients[1], SELINUX_DEFAULT_VERIF, True),
        (self.master, SELINUX_DEFAULT_VERIF, True),
    ],
    user2: [
        (self.clients[0], SELINUX_T3_VERIF, False),
        (self.clients[0], SELINUX_DEFAULT_VERIF, True),
        (self.clients[1], SELINUX_DEFAULT_VERIF, True),
        (self.master, SELINUX_DEFAULT_VERIF, True),
    ],
}

for user, verifications in test_cases.items():
    self._kinit_trust_user(self.master, user, domain)

    for host, expected_selinux_user, expect_match in verifications:
        verify_ssh_selinuxuser_with_krbcred(
            self.master,
            user,
            host,
            expected_selinux_user,
            expect_match=expect_match,
        )
```

**Review by amore17 — CHANGES_REQUESTED** [review]

File: ipatests/test_integration/test_trust_functional.py [hunk 6]
```diff
@@ -3942,6 +3969,601 @@ def test_selinuxusermap_006_hbac_hostgroup_map(self):
...
+                    self.master, user1, self.master, SELINUX_T1_VERIF,
+                    expect_ssh_success=False)
+
+                self._kinit_trust_user(self.master, user2, domain)
+                verify_ssh_selinuxuser_with_krbcred(
+                    self.master, user2, self.clients[0], SELINUX_T1_VERIF,
+                    expect_ssh_success=False)
+                verify_ssh_selinuxuser_with_krbcred(
+                    self.master, user2, self.clients[1], SELINUX_T1_VERIF,
+                    expect_ssh_success=False)
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: You can also improve this with: 
```
# 007: rule7 lets group1 ssh to client1/client2 with staff_u.
# user1 denied on master (outside hostgroups);
# user2 denied everywhere (not in rule7).

test_cases = {
    user1: [
        (self.clients[0], True),
        (self.clients[1], True),
        (self.master, False),
    ],
    user2: [
        (self.clients[0], False),
        (self.clients[1], False),
    ],
}

for user, cases in test_cases.items():
    self._kinit_trust_user(self.master, user, domain)

    for host, expect_success in cases:
        verify_ssh_selinuxuser_with_krbcred(
            self.master,
            user,
            host,
            SELINUX_T1_VERIF,
            expect_ssh_success=expect_success,
        )
```
Keeps the expected outcome (True/False) next to each test case, improving readability.

File: ipatests/test_integration/test_trust_functional.py [hunk 7]
```diff
@@ -3942,6 +3969,601 @@ def test_selinuxusermap_006_hbac_hostgroup_map(self):
...
+                # 008: password SSH from client2 — unconfined_u to master
+                # and client1 for both users.
+                for user in (user1, user2):
+                    self._verify_password_ssh_selinux_from(
+                        self.clients[1], user,
+                        (
+                            (self.master, SELINUX_DEFAULT_VERIF, True),
+                            (self.clients[0], SELINUX_DEFAULT_VERIF, True),
+                        ),
+                    )
```
Comment (amore17) [inline ipatests/test_integration/test_trust_functional.py]: This can be improved with : 
```
# 008: password SSH — both users get unconfined_u on all other hosts.

test_cases = {
    self.clients[0]: (
        (self.master, SELINUX_DEFAULT_VERIF, True),
        (self.clients[1], SELINUX_DEFAULT_VERIF, True),
    ),
    self.clients[1]: (
        (self.master, SELINUX_DEFAULT_VERIF, True),
        (self.clients[0], SELINUX_DEFAULT_VERIF, True),
    ),
}

for source, expectations in test_cases.items():
    for user in (user1, user2):
        self._verify_password_ssh_selinux_from(
            source,
            user,
            expectations,
        )
```

**Comment by amore17** [comment]
@pranav210798 You can remove temp_commit LGTM 
[results](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/7b604c18-79f3-11f1-ae45-fa163e863465/)

**Comment by pranav210798** [comment]
@amore17 Will need a rebase once the PR https://github.com/freeipa/freeipa/pull/8467 is merged. Will need to wait for PR to get merged.

**Comment by pranav210798** [comment]
Rebased and pushed we can ack.

**Comment by flo-renaud** [comment]
master:

* 2b87c8cc6669da0cf7524cd83dabef588d80b618 ipatests: Add AD trust SELinux user map tests (part 3)
